src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/Search/SearchHasVersionValidator.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.Services.Metadata.Catalog.Monitoring
+{
+    public class SearchHasVersionValidator : Validator<SearchEndpoint>
+    {
+        public SearchHasVersionValidator(
+            SearchEndpoint endpoint,
+            ValidatorConfiguration config,
+            ILogger<Validator> logger)
+            : base(endpoint, config, logger)
+        {
+        }
+
+        protected override async Task RunInternalAsync(ValidationContext context)
+        {
+            var searchVisible = await IsVisibleInSearchAsync(context);
+            var databaseState = await GetDatabaseStateAsync(context);
+            var databaseVisible = databaseState == DatabaseState.Listed;
+            if (databaseVisible != searchVisible)
+            {
+                const string listedString = "listed";
+                const string unlistedString = "unlisted";
+
+                throw new MetadataInconsistencyException(
+                    $"Database shows {databaseState.ToString().ToLowerInvariant()}" +
+                    $" but search shows {(searchVisible ? listedString : unlistedString)}.");
+            }
+        }
+
+        private async Task<bool> IsVisibleInSearchAsync(ValidationContext context)
+        {
+            var searchPage = await GetSearchPageForIdAsync(context);
+            var searchItem = searchPage.FirstOrDefault();
+            bool searchListed;
+            if (searchItem != null)
+            {
+                var searchVersions = await searchItem.GetVersionsAsync();
+                searchListed = searchVersions.Any(x => x.Version == context.Package.Version);
+            }
+            else
+            {
+                searchListed = false;
+            }
+
+            return searchListed;
+        }
+
+        private async Task<IEnumerable<IPackageSearchMetadata>> GetSearchPageForIdAsync(ValidationContext context)
+        {
+            return await (Task<IEnumerable<IPackageSearchMetadata>>)context.Cache.GetOrAdd(
```

Comment:
I'm not a huge fan of the free-form "string to object" thing. Having it relatively unstructured like this is both harder to use and opens it up to confusing keys in the future.

Specifically, I'd rather see a "base uri to query endpoint to results" cache just for search results, e.g. `IDictionary<Uri, IDictionary<string, Task<IEnumerable<IPackageSearchMetadata>>>>` or at least a `IDictionary<Uri, Task<IEnumerable<IPackageSearchMetadata>>>` if you are opposed to the multi-level nesting. This could allow you to share a lot of the logic necessary to construct queries across all of the search validations.

For example, you could add something like this to `ValidationContext`:
```
private IDictionary<Uri, IDictionary<string, Task<IEnumerable<IPackageSearchMetadata>>>> _searchQueryCache = new Dictionary<Uri, IDictionary<string, Task<IEnumerable<IPackageSearchMetadata>>>>();

public Task<IEnumerable<IPackageSearchMetadata>> GetPackageFromSearch(Uri baseUri, string endpoint)
{
    if (_searchQueryCache.TryGetValue(baseUri, out var queryCache) && queryCache.TryGetValue(query, out var result))
    {
        return result;
    }
    
    var queryUri = new Uri(Endpoint.BaseUri, endpoint);
    Logger.LogInformation(
        "Searching for package {Id} on {Uri}.",
        Package.Id,
        queryUri.AbsoluteUri);

    var rawSearchResource = new RawSearchResourceV3(context.HttpSourceResource.HttpSource, new[] { queryUri });
    var packageSearchResource = new PackageSearchResourceV3(rawSearchResource);
    return packageSearchResource.SearchAsync(
        searchTerm: $"packageid:{Package.Id}",
        filter: new SearchFilter(includePrerelease: true),
        skip: 0,
        take: 1,
        log: CommonLogger,
        cancellationToken: CancellationToken);
}
```

Reply:
Okay, I was thinking the knowledge should live in the validator but I can do this instead.